kong/pdk/client.lua

```diff
@@ -241,6 +242,50 @@ local function new(self)
     return is_tls and "tls" or "tcp"
   end
 
+  ---
+  -- Returns a table of `authenticated_groups` of the currently authenticated
+  -- consumer.
+  -- @function kong.client.get_authenticated_groups
+  -- @phases access, header_filter, body_filter, log
+  -- @treturn table,nil|err
+  -- The table will have an array part to iterate over, and a hash part where each
+  -- group name is indexed by itself. Eg.
+  -- {
+  --   [1] = "users",
+  --   [2] = "admins",
+  --   users = "users",
+  --   admins = "admins",
+  -- }
+  -- @usage
+  -- local groups, err = kong.client.get_authenticated_groups()
+  -- if not err then
+  --   for i = 1, #groups_to_check do
+  --     if groups[i] == 'group1' then
+  --       return true
+  --     end
+  --   end
+  -- end
+  function _CLIENT.get_authenticated_groups()
+    check_phase(AUTH_AND_LATER)
+
+    local authenticated_groups = ngx.ctx.authenticated_groups
+    if authenticated_groups == nil then
+      return nil
```

Comment:
consideration: you could return an empty table in this case, since this is not necessarily an error condition. The upside would be that error handling becomes easier since you can check the first result, and do not have to check the second one.

Reply:
currently it also does not comply with the doc comments, returning an empty table would.

Reply:
Is there a scenario where `ngx.ctx.authenticated_groups` may be an empty table today? (As in, "authentication was checked and there are no groups" or something?) I'm unfamiliar with the details.

I ask this because looking at the code that uses this API in `kong/plugins/acl/handler.lua`, `nil` (not error) and `{}` would fall into two different paths today.

If the API is made to return `{}` when it currently returns `nil`, will client code need to be adjusted accordingly by testing for `if #authenticated_groups > 0` instead of `if authenticated_groups`?

Another way of putting this is: is there a semantic difference today between `{}` and `nil` when returned by `kong.plugins.acl.groups.get_authenticated_groups`? (As in, for example, "groups were checked and none match" vs. "groups were not checked yet") If so, does this distinction matter?

Reply:
@hishamhm you bring up a good point. The tests do seem to assume that `nil` is semantically different. For instance, currently one test is failing that is checking `nil`ness of groups here:
https://github.com/kong/kong/blob/304c3dfd947011cead2a47a6f6769afb9e98e325/kong/plugins/acl/handler.lua#L72 but merely substituting length of groups everywhere we do authenticated_groups check to check length of array breaks other tests so changing it back unless @bungle has some background he would like to provide.

Reply:
It's also worth noting that now we have `assert`, we no longer need to return err, so `table|nil` for return value is much simpler than `table|nil,err|nil`.